### PR TITLE
Merge disruption

### PR DESCRIPTION
**Pull request by antoine-de**
Handling of Chaos disruptions in navitia.

still some work to be done (especially in jormungandr and we need to test the AT connector)

branch from #634 merged into dev

### CODE REVIEW

File: source/type/realtime.proto [hunk 1]
```diff
@@ -6,30 +6,6 @@ message TCObject{
     optional pbnavitia.NavitiaType object_type = 2;
 }
 
-message LocalizedMessage{
```
Comment (kinnou02) [inline source/type/realtime.proto:9]: it's used in sindri and at_connector no?

Comment (antoine-de) [inline source/type/realtime.proto:9]: doesn't seems no (nothing with git grep LocalizedMessage)

File: source/disruption/tests/disruption_test.cpp [hunk 2]
```diff
@@ -57,97 +59,155 @@ start_application_daily_hour = 08h40
...
+BOOST_GLOBAL_FIXTURE( logger_initialized )
+
 namespace pt = boost::posix_time;
-class Params{
+using navitia::type::new_disruption::Impact;
+using navitia::type::new_disruption::PtObj;
+using navitia::type::new_disruption::Disruption;
+
+enum class ChaosType {
+    //we copy the chaos type because we don't want chaos dependencies here
```
Comment (TeXitoi) [inline source/disruption/tests/disruption_test.cpp]: Why we don't want chaos dependencies in a test? who care?

File: source/ed/at_perturbation.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    //intersection de la period ou se déroule l'action et de la periode de validité de la perturbation
+    pt::time_period common_period = period.intersection(application_period);
+
+    //si la periode commune est null, l'impact n'est pas valide
+    if(common_period.is_null()){
+        return false;
+    }
+
+    //on doit travailler jour par jour
```
Comment (TeXitoi) [inline source/ed/at_perturbation.cpp]: french comments? ;)

File: source/ed/connectors/messages_connector.cpp [hunk 4]
```diff
@@ -35,12 +35,47 @@ www.navitia.io
...
+    }
+
+    auto it = disruptions.severities.find(name);
+
+    if (it != disruptions.severities.end()) {
+        // we return the weak_ptr if it is still alive, else we'll add a new one
+        if (auto res = it->second.lock()) return res;
+    }
+
+    boost::shared_ptr<Severity> severity = boost::make_shared<Severity>();
```
Comment (TeXitoi) [inline source/ed/connectors/messages_connector.cpp]: auto is great for useless repetition of type like this

File: source/ed/connectors/messages_connector.cpp [hunk 5]
```diff
@@ -69,114 +104,71 @@ std::map<std::string, boost::shared_ptr<navitia::type::Message>> load_messages(
...
-            cursor["uri"].to(current_uri);
-            //on construit le message
-            message = boost::make_shared<navitia::type::Message>();
-            message->uri = current_uri;
-            cursor["object_uri"].to(message->object_uri);
-
-            message->object_type = static_cast<navitia::type::Type_e>(
-                    cursor["object_type_id"].as<int>());
+    for (auto cursor = result.begin(); cursor != result.end(); ++cursor) {
+        //we can have several language handled by the database, so we create one message for each
```
Comment (TeXitoi) [inline source/ed/connectors/messages_connector.cpp]: I don't understand this comment. where is the loop adding a message for each language?

File: source/ed/connectors/messages_connector.cpp [hunk 6]
```diff
@@ -35,12 +35,47 @@ www.navitia.io
...
+    if (it != disruptions.severities.end()) {
+        // we return the weak_ptr if it is still alive, else we'll add a new one
+        if (auto res = it->second.lock()) return res;
+    }
+
+    boost::shared_ptr<Severity> severity = boost::make_shared<Severity>();
+    disruptions.severities.insert({name, severity});
+
+    severity->uri = name;
+    severity->effect = Effect::UNKNOWN_EFFECT; //is it right ?
```
Comment (kinnou02) [inline source/ed/connectors/messages_connector.cpp]: yes

File: source/type/realtime.proto (hunk 1 above)
Comment (kinnou02) [inline source/type/realtime.proto:9]: sindri/sindri/saver/message.py:126
connectors/connectors/at/selector/atreader.py:253

File: source/kraken/fill_disruption_from_chaos.cpp [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    impact->uri = chaos_impact.id();
+    impact->created_at = from_posix(chaos_impact.created_at());
+    impact->updated_at = from_posix(chaos_impact.updated_at());
+    for (const auto& chaos_ap: chaos_impact.application_periods()) {
+        impact->application_periods.emplace_back(from_posix(chaos_ap.start()), from_posix(chaos_ap.end()));
+    }
+    impact->severity = make_severity(chaos_impact.severity(), holder);
+    impact->informed_entities = make_pt_objects(chaos_impact.informed_entities(), pt_data, impact);
+    for (const auto& chaos_message: chaos_impact.messages()) {
+        impact->messages.push_back({
```
Comment (kinnou02) [inline source/kraken/fill_disruption_from_chaos.cpp:174]: I don't see where we handle the channel

File: source/kraken/maintenance_worker.cpp [hunk 8]
```diff
@@ -94,15 +99,22 @@ void MaintenanceWorker::handle_rt(AmqpClient::Envelope::ptr_t envelope){
...
         return;
     }
+    boost::shared_ptr<nt::Data> data;
     for(const auto& entity: feed_message.entity()){
         if(entity.HasExtension(chaos::disruption)){
             LOG4CPLUS_WARN(logger, "has_extension");
-            data_manager.apply_disruptions(0);
+            if (!data) {
+                data = data_manager.get_data_clone();
+                data->last_rt_data_loaded = pt::microsec_clock::local_time();
```
Comment (kinnou02) [inline source/kraken/maintenance_worker.cpp]: local or utc?

File: source/kraken/fill_disruption_from_chaos.cpp (hunk 7 above)
Comment (antoine-de) [inline source/kraken/fill_disruption_from_chaos.cpp:174]: nop, they're not handled for the moment.
we just need to read them or do we need some filters to ?

**Comment by TeXitoi** [comment]
OK for me.

**Pull request merged**